controllers/storageautoscaler/storageautoscaler_controller.go

```diff
@@ -0,0 +1,286 @@
+func (r *StorageAutoscalerReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	r.Log.Info("reconciling storage autoscaler", "namespace", request.Namespace, "name", request.Name)
+
+	// list the storage autoscaler reconciler object
+	storageAutoScaler := &ocsv1.StorageAutoScaler{}
+	err := r.Get(ctx, client.ObjectKey{Namespace: request.Namespace, Name: request.Name}, storageAutoScaler)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("storage autoscaler not found for reconcile request", "namespace", request.Namespace, "name", request.Name)
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		r.Log.Error(err, "failed to get storage autoscaler")
+		return reconcile.Result{}, err
+	}
+
+	// if no expansion is in-progress update the phase to "not-started"
+	if storageAutoScaler.Status.Phase == "" {
+		originalStorageAutoScaler := storageAutoScaler.DeepCopy()
+		storageAutoScaler.Status.Phase = "NotStarted"
+		err := r.updateStatus(ctx, storageAutoScaler, originalStorageAutoScaler)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// if the expansion is InProgress or Failed, verify the scaling
+	if storageAutoScaler.Status.Phase == "InProgress" || storageAutoScaler.Status.Phase == "Failed" {
+		return r.verifyScaling(ctx, storageAutoScaler)
+	}
+
+	// if the last expansion is succeeded 10 minutes ago, do not scale
+	if storageAutoScaler.Status.LastExpansion != nil {
+		if time.Since(storageAutoScaler.Status.LastExpansion.CompletionTime.Time).Seconds() < 600 {
+			return reconcile.Result{}, nil
+		}
+	}
+
+	// list the storagecluster
+	storageCluster := &ocsv1.StorageCluster{}
+	storageClusterName := storageAutoScaler.Spec.StorageCluster.Name
+	err = r.Get(ctx, client.ObjectKey{Name: storageClusterName, Namespace: request.Namespace}, storageCluster)
+	if err != nil {
+		r.Log.Error(err, "failed to get storage cluster")
+		return reconcile.Result{}, err
+	}
+
+	// list the storagecluster devicset storageclass
+	err = r.detectLsoStorageclass(ctx, storageCluster)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// check if the ceph cluster is healthy
+	notHealthy, err := r.checkIfCephClusterIsNotHealthy(ctx, storageCluster)
+	if err != nil {
+		r.Log.Error(err, "failed to check if ceph cluster is healthy")
+		return reconcile.Result{}, err
+	}
+	if notHealthy {
+		r.Log.Info("ceph cluster is not healthy cannot scale")
+		return reconcile.Result{}, nil
+	}
+
+	// get the osd usage from the sync map
+	usage, ok := r.SyncMap.Load(storageAutoScaler.Spec.DeviceClass)
+	if !ok {
+		err = fmt.Errorf("osd usage not found for device class, device class: %s", storageAutoScaler.Spec.DeviceClass)
+		r.Log.Error(err, "device class not found in sync map")
+		return reconcile.Result{}, err
```

Comment:
@travisn  should we return `error: nil` till the sync map is not updated? Otherwise controller will give an error for 10 mins till the scraper doesn't update the sync map.

Reply:
If we return the error, the reconcile will continue to be re-queued and keep getting errors, right? When would this happen? Perhaps at first startup before the prometheus goroutine has run? If it's a normal condition, then I agree we should just log that we are skipping the expansion evaluation and return nil.

Reply:
Yes if needed scarper will resend the request, it mostly needed to don't fail the controller at startup when the cr is just created and the scraper is not ran.